gadget/install/install.go

```diff
@@ -247,24 +266,106 @@ func Run(model gadget.Model, gadgetRoot, kernelRoot, bootDevice string, options
 			ExpectedStructureEncryption: partsEncrypted,
 		},
 	}
-	allVolTraits, err := gadget.AllDiskVolumeDeviceTraits(allLaidOutVols, optsPerVol)
-	if err != nil {
+	// save the traits to ubuntu-data host and optionally to ubuntu-save if it exists
+	if err := saveStorageTraits(allLaidOutVols, optsPerVol, hasSavePartition); err != nil {
 		return nil, err
 	}
 
-	// save the traits to ubuntu-data host
-	if err := gadget.SaveDiskVolumesDeviceTraits(dirs.SnapDeviceDirUnder(boot.InstallHostWritableDir), allVolTraits); err != nil {
-		return nil, fmt.Errorf("cannot save disk to volume device traits: %v", err)
+	return &InstalledSystemSideData{
+		KeysForRoles: keysForRoles,
+	}, nil
+}
+
+func FactoryReset(model gadget.Model, gadgetRoot, kernelRoot, bootDevice string, options Options, observer gadget.ContentObserver, perfTimings timings.Measurer) (*InstalledSystemSideData, error) {
+	logger.Noticef("performing factory reset on an installed system")
+	logger.Noticef("        gadget data from: %v", gadgetRoot)
+	logger.Noticef("        encryption: %v", options.EncryptionType)
+	if gadgetRoot == "" {
+		return nil, fmt.Errorf("cannot use empty gadget root directory")
 	}
 
-	// and also to ubuntu-save if it exists
-	if hasSavePartition {
-		if err := gadget.SaveDiskVolumesDeviceTraits(boot.InstallHostDeviceSaveDir, allVolTraits); err != nil {
-			return nil, fmt.Errorf("cannot save disk to volume device traits: %v", err)
+	if model.Grade() == asserts.ModelGradeUnset {
+		return nil, fmt.Errorf("cannot run factory-reset mode on non-UC20+ system")
+	}
+
+	if options.EncryptionType != secboot.EncryptionTypeNone {
+		return nil, fmt.Errorf("factory-reset on encrypted system is unsupported")
+	}
+
+	laidOutBootVol, allLaidOutVols, err := gadget.LaidOutVolumesFromGadget(gadgetRoot, kernelRoot, model)
+	if err != nil {
+		return nil, fmt.Errorf("cannot layout volumes: %v", err)
+	}
+	// TODO: resolve content paths from gadget here
+
+	// XXX: the only situation where auto-detect is not desired is
+	//      in (spread) testing - consider to remove forcing a device
+	//
+	// auto-detect device if no device is forced
```

Comment:
is this comment relevant at all for FactoryReset? do we pass always pass or never pass bootDevice?

Reply:
There's uc20-create-partitions tool which passes the boot device. I planned to extend this to maybe also execute the factory reset path.

Reply:
so we have an XXX that we can't really do anything about,  we probably should just a keep a comment that setting bootDevice is done in spread tests without the XXX wishful bits, same in Run